sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/interpreter/operator/UdafImpl.java

```diff
@@ -58,9 +58,12 @@ public String getName() {
           }
 
           public RelDataType getType(RelDataTypeFactory typeFactory) {
-            //the first generic type of CombineFn is the input type.
-            ParameterizedType parameterizedType = (ParameterizedType) combineFn.getClass()
-                .getGenericSuperclass();
+            Class clazz = combineFn.getClass();
+            while (!(clazz.getGenericSuperclass() instanceof ParameterizedType)) {
```

Comment:
This finds only first generic superclass, would it fail if there's another generic type between your UDAF subclass and `CombineFn` (imagine `YourUDAF extends YourGenericUDAF<SomeType> extends CombineFn<Integer, Integer, Integer>`). Should this loop check whether the `clazz` is also actually the `CombineFn` and not just some generic type?

And I think it's better to throw if `ParametrizedType` is not found, this implementation doesn't work without it, so it's better to make it explicit and not rely on cast exceptions, NPEs, etc.

Reply:
You're right. I didn't think that through fully... it definitely makes sense to look specifically for CombineFn. So would it be sufficient to loop through superclasses looking for CombineFn && ParameterizedType (since a raw extension of CombineFn would be invalid, right?). Then, if we encounter null, through an appropriate exception?

Reply:
Sounds right. I don't think it will ever be null though. It's know for sure it's a subclass of `CombineFn`. Sounds like the code should throw when the loop reaches the `CombineFn` which is not a `ParametrizedType` (not sure if this can happen in reality, probably worth adding a test specifically for this).

Reply:
ah, right, we already know it's a CombineFn. I can add a test case for that. But I think if it's not possible to have a CombineFn that is not also a ParameterizedType in this point of the code, then we don't need to throw an exception, right? That would indicate a bug in Beam SQL. It's not something the user should ever encounter. Does that make sense?

Reply:
If it is impossible to use a non-`ParametrizedType` `CombineFn`, then I agree, there's probably no real benefit in having an extra throw.

Reply:
I verified: I can provide a raw CombineFn subclass instance, and it makes it to this code and fails with: 
```Caused by: java.lang.ClassCastException: sun.reflect.generics.reflectiveObjects.TypeVariableImpl cannot be cast to java.lang.Class
	at org.apache.beam.sdk.extensions.sql.impl.interpreter.operator.UdafImpl$1.getType(UdafImpl.java:68)```

So this method should throw an exception indicating subclasses of CombineFn must be parameterized in order to be used as UDAFs? It seems like that's sufficient to catch all misuse

Reply:
It seems to me that maybe that validation should happen upstream though...

Reply:
Yes, I think it will be sufficient for the current UDAF implementation

Reply:
If the `UdafImpl` asserts something about its input (e.g. that it is a `ParametrizedType`), then it should perform its own validation. Technically no one stops people from using it outside of its current use case. 

Right now it has to carry the `CombineFn` for later use, so it makes sense to validate it internally. But a builder for this class would be a better place for this validation.

Reply:
Yeah, I meant upstream of getType(), not upstream of the class. But anyway, sounds good. Thanks for the discussion.